### ISSUE TITLE
Interop metrics: HTTP fetching O(100k) test files is infeasible

### ISSUE THREAD

**Issue by lukebjerring** [web-platform-tests/results-collection#454]
Runs produces ~22k individual test-file blobs - fetching all of these for 4 runs, individually is infeasibly slow, as (even if HTTP wasn't a bottleneck) the Cloud Storage API rate-limits requests.

While it is possible to create a client that batches these requests, it is a workaround at best.
It is apparent that the choice to shard the results hinges on the impact of an excessively large aggregated file on the wpt.fyi UI. However, having shards, and having aggregated files, is not mutually exclusive. We need to cover the use-case of processing all results (not just a per-file summary) across multiple browsers.

**Comment by rwaldron** [comment]
As far as I can tell, https://github.com/w3c/wptdashboard/blob/master/metrics/run/collect_metrics.go orchestrates with https://github.com/w3c/wptdashboard/blob/master/metrics/compute/compute.go to run some addition functions. The same interop metrics information could be "computed" from the existing summary results data files that are fetched based on the response from https://wpt.fyi/api/runs?sha=latest, and the subtests results files that are fetched at the individual test file level.

**Comment by lukebjerring** [comment]
Summary data does not include the counts of subtests; as-is, metrics does - are you proposing that the metrics computation be changed to match that (compute summaries from summaries, and individual files/subtests from individual test results)?
Sounds reasonable, I think I suggested as much - @mdittmer, anything we're missing?

**Comment by foolip** [comment]
I think the current state is that we take the output of `wpt run` and from that produce the summary files as well as tens of thousands of split files. However the stuff that's served to the web property evolves, I think we could solve this by simply also commit to datastore the original pre-split files. The metrics computation code would then have to load just 4 files. Would that work?

**Comment by mdittmer** [comment]
That seems like a good idea to me. I have two thoughts:

1. Upload the large pre-split file to GCS, not Datastore;
2. We need to watch out for what the implications for a merged file will be for infra (details below).

### Caveats for one-big-file-per-run solution

If we are (and I think we are, right @mattl?) starting to split runs up across multiple machines, or multiple "tries" if things break part way through, care is required to ensure that the one-big-file-per-run contains all results.

One potential design is to have the bot that "finishes the last tests for a run", either:

1. Gather results and produce the large file, or
2. Broadcast a signal that tells another piece of infrastructure to do this.

Unfortunately, whatever component puts the pieces back together will face the same issue that we're seeing in Golang.

Another solution would be to use a field-level-update-friendly mechanism like Firestore, have bots running tests send updates that constitute their (potentially partial) test results to this mechanism, and let it deal with merging things together. The risk with this solution is that such mechanisms often have per-entity/document quotas that are really slow (because of the locking required), on the order of 1 update per entity/document per second. This is fine if the same bot needs to send updates on retry, but not fine if we move to a highly parallel world where a whole fleet of bots push their small chunk of test results A.S.A.P..

**Comment by foolip** [comment]
> Upload the large pre-split file to GCS, not Datastore

I didn't know which words I was using, I mean you you said :)

**Comment by mariestaver** [comment]
Making this change is out of scope for our plan this quarter.  We can leave the issue open to track discussion (please continue it if you like), but in all honesty this work can't even be evaluated for possible implementation until the runner & infrastructure are in better shape.

*I'll create a label* to track this and other issues and ideas that should be considered for possible inclusion in the next (or a future) quarter.  We'll update all the issues with this label at the end of the current quarter with our findings and whether they can be part of the next quarter's plan.

(I'm not sure what to call the label yet, so please bear with me while I figure out what's a less-confusing name than 'backlog' etc.  I'll add it as soon as the right thing occurs to me.)

**Comment by mariestaver** [comment]
Update:  It looks like the existing "priority:backlog" label is rarely/never used, so I will adopt it for this use.  Please expect to see an update on those issues near the beginning of each new quarter.  Thanks!

**Comment by jugglinmike** [comment]
Although @lukebjerring originally raised this issue with metrics calculation in mind, the recent changes to project organization/ownership make this topic even more important. We should formally commit to a data format that can remain stable into the future.

In particular, this means re-thinking any design decisions that were informed by the needs of the current iteration of https://wpt.fyi. In addition to supporting metrics calculations, a generalized data format will help other folks use the data as a resource. And formally committing to stability will make comparing historic results easier.

Separately, I've been interested in promoting transparency and reproducibility by extending these reports with metadata describing the conditions under which the results were collected. We can defer decisions on the metadata for the time being; for now, I'd just like to make sure we reserve a place for it in the data model.

Currently, we produce reports that are composed of many JSON-formatted and gzip-encoded files: a single "summary" file and many test results files (one for each test in WPT--about 25,000 at the time of this writing). Here's a general sketch of their contents:

- Summary file

      {
        "<string:test name>": [
          "<int:number of passing sub-tests + 1>",
          "<int:total number of sub-tests + 1>"
        ],

        /* about 25,000 more entries */
      }

- Results file

      {
        "test": "<string:test name>",
        "message": "<string or null>",
        "status": "<string>",
        "subtests": [
          {
            "message": "<string or null>",
            "name": "<string>",
            "status": "<string>"
          },

          /* any number of additional items */
        ]
      }

Note that the results files match the format of the data emitted by the WPT CLI (via the `--log-wptreport` flag).

Here's how they are laid out in a file system:

    (results directory)/
    |- <browser-identifier>-summary.json.gz
    `- <browser-identifier>/
       |- <path>/<to>/<WPT>/<test>
       |
       | (about 25,000 more files)
       |
       ` <path>/<to>/<WPT>/<test>

In line with [Philip's suggestion above](https://github.com/w3c/wptdashboard/issues/454#issuecomment-365700801), I'd like to instead publish a single file: the `--log-wptreport` produced by the WPT CLI. (On the results collection end, this will actually be created by consolidating a series of partial results, but that detail won't be apparent to consumers.) I'd also like to extend it with some additional information: the date of collection, the git revision of WPT, and a (for now) unstructured collection of environment data:

    {
      "date": "<string>",
      "revision": "<string>",
      "environment": {},
      "tests": [
        {
          "test": "<string:test name>",
          "message": "<string or null>",
          "status": "<string>",
          "subtests": [
            {
              "message": "<string or null>",
              "name": "<string>",
              "status": "<string>"
            }
          ]
        },

        /* about 25,000 more entries */
      ]
    }

This file will be quite large (about 150 megabytes), so I think it makes sense to compress it. gzip will get the file down to 12 megabytes, which may be good enough to preclude the need for more specialized algorithms like BSON or MessagePack.

**In summary**: this change would mean that the WPT Dashboard application will have a new responsibility: creating files/objects from a single 12 megabyte gzipped JSON file. @lukebjerring can you let me know if this will work for you?  If so, I'll create a new bucket for the new format and publish both formats.  That will allow you to switch to the new format at your own pace.

**Comment by jugglinmike** [comment]
I just found an mistake in my original example: the proposed results format used a top-level key named "tests" instead of "results". I've updated the example accordingly; sorry for the noise!

**Comment by lukebjerring** [comment]
Our intention is to standardize the format that the upload endpoint accepts, rather than the format in which the results are stored in cloud storage.
This issue (metrics storage format changes) will be deferred until we've changed the runner to upload non-summarized results.

We should create a more clearly named issue for the runner (Upload a single, unsummarized results report) blocked on the webapp work (Support upload of a single, unsummarized results report). Your comments on gzipping mirror my thoughts and apply to the latter issue.

Also, don't get me wrong, we should standardize the storage format too - but facilitating 3rd party runners needs a stable endpoint, so the storage format can iterate independently, while we figure out implications on the loading of the UI.

**Comment by jugglinmike** [comment]
Sure--separate issue here: https://github.com/w3c/wptdashboard/issues/527
